File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_1_100/Sol11.java
/*
 * Copyright NetEase, Inc. ALl rights reserved.
 */

package com.goinhn.thorwcode.leetcode.question._1_100;

/**
 * 11.盛最多水的容器
 *
 * @author fangwenzhan
 * @date 2022/10/4
 */
public class Sol11 {

    /**
     * 贪心算法
     * time:O(n)
     * space:O(1)
     */
    private static class Solution1 {

        public int maxArea(int[] height) {
            int left = 0;
            int right = height.length - 1;
            int ans = Integer.MIN_VALUE;
            while (left < right) {
                int now = (right - left) * (Math.min(height[left], height[right]));
                ans = Math.max(now, ans);
                if (height[left] < height[right]) {
                    left++;
                } else {
                    right--;
                }
            }

            return ans;
        }

    }

}

File: leetcode_go/question/_1_100/_11/sol_11.go
// Package _11 盛最多水的容器
// @author goinhn
// @data 2022/10/07
package _11

// 贪心
// time:O(n)
// space:O(1)
func maxArea(height []int) int {
	left := 0
	right := len(height) - 1
	ans := 0
	for left < right {
		hi := 0
		length := right - left
		if height[left] < height[right] {
			hi = height[left]
			left++
		} else {
			hi = height[right]
			right--
		}

		now := length * hi

		if ans < now {
			ans = now
		}
	}

	return ans
}
